okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/FastFallbackExchangeFinder.kt

```diff
@@ -35,132 +35,139 @@ internal class FastFallbackExchangeFinder(
 ) {
   private val connectDelayMillis = 250L
 
-  /** Plans currently being connected, and that will later be added to [connectResults]. */
-  private var connectsInFlight = CopyOnWriteArrayList<Plan>()
+  /**
+   * Plans currently being connected, and that will later be added to [connectResults]. This is
+   * mutated by the call thread only. If is accessed by background connect threads.
+   */
+  private val tcpConnectsInFlight = CopyOnWriteArrayList<Plan>()
+
+  /** Plans we started, and canceled, because another plan won a race. */
+  private val deferredPlans = ArrayDeque<Plan>()
 
   /**
    * Results are posted here as they occur. The find job is done when either one plan completes
    * successfully or all plans fail.
    */
   private val connectResults = taskRunner.backend.decorate(LinkedBlockingDeque<ConnectResult>())
 
-  /** Exceptions accumulate here. */
-  private var firstException: IOException? = null
-
   /** True until we've launched all the connects we'll ever launch. */
-  private var morePlansExist = true
+  private var routePlannerHasMoreRoutes = true
 
   fun find(): RealConnection {
+    var firstException: IOException? = null
     try {
-      while (morePlansExist || connectsInFlight.isNotEmpty()) {
+      while (routePlannerHasMoreRoutes ||
+        tcpConnectsInFlight.isNotEmpty() ||
+        deferredPlans.isNotEmpty()
+      ) {
         if (routePlanner.isCanceled()) throw IOException("Canceled")
 
-        launchConnect()
+        launchTcpConnect()
```

Comment:
Do you always want to start a new connection on each go around of the loop? If you already have 2-3 on the go, you may not have waited the minimum 100ms "Minimum Connection Attempt Delay", and may either want to delay or check if other plans have tcp connected also.

Reply:
Ooooh this is a great find. Lemme write a test case for that.